### ISSUE TITLE
Spout Input as Sprites!

### ISSUE THREAD

**Issue by OfficialIncubo** [OfficialIncubo/BeatDrop-Music-Visualizer#26]
Hey! It would be cool if this feature will be integrated in! This will be the Mini-NestDrop!

**The usage:**
If you want to invoke a sprite as a Spout Input, then beatdrop_img.ini will be necessary to access images/Spout Inputs.

Add `Spout/SpoutInputName` before `img=` on any digit headers you want, so it should be able to access it.

For example: The Spout OBS Plug-in Name that I route a YouTube Video is **YouTubeSpoutActivator**, so it should be `img=Spout/YouTubeSpoutActivator` in the [img11] header on beatdrop_img.ini.
If I try to change to Sprite Mode and type 11 (e.g. K11), it will invoke a Spout Input Sprite routed rrom OBS.

Hey, @leadedge, a.k.a. Lynn Jarvis, you are the one who made the Spout Output Integration of BeatDrop and I am the one who improves this project. Will it be possible to add a Spout Sprite Input functionality like the example that I have mentioned?

**Comment by leadedge** [comment]
You would have to set up a receiver or multiple receivers somehow, but it should be possible. I can't remember whether you are using OpenGL or DirectX but The example receiver source code for [OpenGL](https://github.com/leadedge/Spout2/tree/master/SPOUTSDK/Examples/ofSpoutExample/Source/Receiver/Graphics) or SpoutDX for [DirectX](https://github.com/leadedge/Spout2/blob/master/SPOUTSDK/SpoutDirectX/SpoutDX/Tutorial07/Tutorial07.cpp) and [Windows](https://github.com/leadedge/Spout2/tree/master/SPOUTSDK/SpoutDirectX/SpoutDX/Windows/Receiver) applications would be a reference.

**Comment by OfficialIncubo** [comment]
> I can't remember whether you are using OpenGL or DirectX but The example receiver source code for [OpenGL](https://github.com/leadedge/Spout2/tree/master/SPOUTSDK/Examples/ofSpoutExample/Source/Receiver/Graphics) or SpoutDX for [DirectX](https://github.com/leadedge/Spout2/blob/master/SPOUTSDK/SpoutDirectX/SpoutDX/Tutorial07/Tutorial07.cpp) and [Windows](https://github.com/leadedge/Spout2/tree/master/SPOUTSDK/SpoutDirectX/SpoutDX/Windows/Receiver) applications would be a reference.

Yes! BeatDrop uses SpoutDX for performance improvement, also for Spout Output.
Where can I paste the entire code of "Tutorial07.cpp"? In plugin.cpp, I think there's a Sprite Activation, so what should I enable support for Spout Input Sprites when it detects `img=Spout/InputName`?

**Comment by leadedge** [comment]
Tutorial07 is a complete example from Microsoft, modified for Spout functions, so it's not possible to copy/paste the entire thing. 

Unfortunately, I forgot that BeatDrop uses DirectX 9, so [this](https://github.com/leadedge/Spout2/tree/master/SPOUTSDK/SpoutDirectX/SpoutDX/SpoutDX9/Tut02_Vertices_Receiver) is a relevant example using SpoutDX9. `ReceiveDX9Texture` will give a texture to work with. I am sorry that I can't assist with BeatDrop code.

**Comment by OfficialIncubo** [comment]
But hey, @leadedge! You can take a look in `plugin.cpp` -> `bool CPlugin::LaunchSprite` -> part of the comment starting from `// 1. read in image filename`. This is a functionality that gets an image filename, which it would have ability to read the `Spout\` prefix then the sender name, so it should have to receive the Spout Texture as a sprite and it will be effected by the init and per-frame code from the image ini file, like NestDrop.

You can also look in `texmgr.cpp` -> `int texmgr::LoadTex`. As you can see that it's an algorithm that it loads a sprite or a texture.

A question: should it work if I try to receive a selected Spout Sender Texture (ex. YouTubeSpoutActivator) read from the image ini file, for example `ReceiveDX9Texture(YouTubeSpoutActivator)`?

**Comment by leadedge** [comment]
There is no support in SpoutDX9 for image files or pixel buffers. You might be able to import the images to a surface with [D3DXLoadSurfaceFromFile](https://learn.microsoft.com/en-us/windows/win32/direct3d9/d3dxloadsurfacefromfile).

**Comment by OfficialIncubo** [comment]
Oh well. I called the function in a "wrong" way. So in [Milkwave](https://github.com/IkeC/Milkwave), it easily calls `bool received = m_pSpoutReceiver->ReceiveDX9Texture(m_pSpoutInputTexture);`, which `m_pSpoutInputTexture` is a `IDirect3DTexture9*` and `m_pSpoutReceiver` is a `spoutDX9*`, called from a local SpoutDX9 file.
Maybe we should get it work in a sprite handler.

Here's a function about UpdateSpoutInputTexture: https://github.com/IkeC/Milkwave/commit/9add1e68ddbd4cd5d332b1eb0999de5a7757a272#diff-88c420fdd8b4123c89faa0f89c50509852573830efd5fcdfeec802a3a4fbff61R45-R144

**Comment by OfficialIncubo** [comment]
You can have a look at the WIP branch: https://github.com/OfficialIncubo/BeatDrop-Music-Visualizer/tree/gif-video-spout-support

Let me know if it works or it's good for you.

**Comment by leadedge** [comment]
Is there a binary I can test with?
Also let me know what to do to test this feature.

**Comment by OfficialIncubo** [comment]
Sadly, I've messed up some code and/or build workflows. That's why it got some errors.

Please try the latest beta test package: https://drive.google.com/file/d/1jVTvD34TgCFCejtEPhkAF0nvkf8ZJ_9-/view?usp=drivesdk

How to use a Spout Input Sprite?
1. Open `beatdrop_img.ini` and find an empty sprite digit (e.g. [imgXX]).
2. Feed an image path using Spout\ or Spout/ prefixes and the sender name (e.g. img=Spout\BeatDrop; both prefixes work)
3. Write some sprite code as an experiment or what you want.
4. The mapped sprite number is 25 ([img25]), for example. Simply press K to change to sprite mode (Y for message/text mode), introduce 2 digits using numbers or numpad (ex. press K, then type 25), then it will automatically invoke a Spout sprite with the effects based on sprite code (if possible). If you introduced a digit in a wrong way, simply use * (or Shift + 8) to reset digits what you've typed.
5. Et voila. Enjoy!

Also test:
- GIF and Video sprites: pass `.gif`, `.mp4` or any video format extensions to `img=` using `beatdrop_img.ini`.

- GIF and Video textures: features an ecosystem that samples animated textures dynamically (e.g. sample an image or sample a GIF/Video; with `sampler_psychedelic` or `sampler_astronaut_swim`).

- Test if it doesn't crash by trying to load a GIF or Video texture when starting up BeatDrop (in my tests, it didn't crash)

**Comment by OfficialIncubo** [comment]
Yep. All builds have passed, tested and working good. 🚀
Did a little fix about CRT link conflict: https://github.com/OfficialIncubo/BeatDrop-Music-Visualizer/commit/8f0700efa08cc625a85b8e9c13c24694028053b4

**Comment by OfficialIncubo** [comment]
One thing that: the video frame parsing is sometimes slow or lazy. This needs a refinement.
